src/main/scala/com/fulcrumgenomics/util/DemuxFastqs.scala

```diff
@@ -0,0 +1,579 @@
+  * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+  * POSSIBILITY OF SUCH DAMAGE.
+  */
+package com.fulcrumgenomics.util
+
+import java.io.File
+import java.nio.file.Path
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, JeanLucTool}
+import com.fulcrumgenomics.util.miseq.{Sample, SampleBarcode, SampleSheet}
+import dagr.commons.CommonsDef.{DirPath, FilePath, PathToFastq, _}
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.fastq.{FastqReader, FastqRecord}
+import htsjdk.samtools.metrics.MetricsFile
+import htsjdk.samtools.util._
+import picard.illumina.ExtractIlluminaBarcodes
+import picard.illumina.parser.ReadStructure
+import picard.util.IlluminaUtil
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+object DemuxFastqs {
+  val UnmatchedSampleId: String = "unmatched"
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  // TODO: duplicated but private in picard.sam.FastqToSam so make it public and static there, and use it here
+  /** Creates the read name to put into a SAM record from a FASTQ record. */
+  private def getReadName(fastqHeader: String, paired: Boolean, stripUnpairedMateNumber: Boolean): String = {
+    val idx: Int = fastqHeader.indexOf(" ")
+    val readName: String = if (idx == -1) fastqHeader else fastqHeader.substring(0, idx)
+    if (stripUnpairedMateNumber && !paired && (readName.endsWith("/1") || readName.endsWith("/2"))) {
+      readName.substring(0, readName.length - 2)
+    }
+    else {
+      readName
+    }
+  }
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call. Observed
+    * qualities may not be null. */
+  private[util] def countMismatches(observedBases: Array[Byte], observedQualities: Array[Byte], expectedBases: Array[Byte], minimumBaseQuality: Int): Int = {
+    observedBases.zip(observedQualities).zip(expectedBases).count {
+      case ((observedBase, observedQuality), expectedBase) =>
+         (!SequenceUtil.isNoCall(expectedBase)) &&
+           (!SequenceUtil.basesEqual(observedBase, expectedBase) || (observedQuality < minimumBaseQuality))
+    }
+  }
+
+  /** Gets all the bases (concatenated) for the given cycle indexes (1-based), ignoring cycles not after the end
+    * of `bases`.  The cycles should be in ascending order. */
+  private def getBasesAtCycles(bases: Array[Byte], cycles: Array[Int]): Array[Byte] = {
+    //  ignore cycles past the end of `bases`.
+    cycles.filter(cycle => cycle - 1 < bases.length).map(cycle => bases(cycle-1))
```

Comment:
the filter could be just `(_ <= bases.length)`, which I think is actually clearer.